utils/filecache/cache.go

```diff
@@ -0,0 +1,227 @@
+package filecache
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	fs "github.com/ARM-software/golang-utils/utils/filesystem"
+	"github.com/ARM-software/golang-utils/utils/parallelisation"
+)
+
+type CacheEntry struct {
+	path       string
+	expiration time.Time
+	ttl        time.Duration
+}
+
+type Cache struct {
+	entries       map[string]*CacheEntry
+	fs            fs.FS
+	cfg           *FileCacheConfig
+	gcCancelStore *parallelisation.CancelFunctionStore
+	entryProvider IEntryProvider
+	mu            sync.RWMutex
+	closed        bool
+}
+
+func (c *Cache) gc(ctx context.Context, _ time.Time) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	now := time.Now()
+	for id, entry := range c.entries {
+		if now.After(entry.expiration) {
+			if err := c.fs.RemoveWithContext(ctx, entry.path); err == nil {
+				delete(c.entries, id)
+			}
+		}
+	}
+}
```

Comment:
There is https://pkg.go.dev/maps#DeleteFunc that can get rid of all this nesting
```suggestion
	maps.DeleteFunc(c.entries, func(id string, entry *CacheEntry) bool {
		return now.After(entry.expiration) && c.fs.RemoveWithContext(ctx, entry.path) == nil
	})
}
```